Add DeviceType property to distinguish between Mice and Keyboards

K1ngjulien opened the pull request:
for now i have implemented parsing of the .device file DeviceType property.
Any feedback would be appreciated.


Now I need to find out how to expose that property over the DBus. Help with that would be much appreciated.

FFY00 commented on src/libratbag-data.c:
```diff
@@ -53,6 +53,12 @@ enum driver {
+enum ratbag_device_type {
+	MOUSE = 0,
+	KEYBOARD,
+	TYPE_NONE
```
```c
enum ratbag_device_type {
	NONE = 0,
	MOUSE,
	KEYBOARD,
}
```

K1ngjulien replied:
I get a compiler error because driver also has `NONE` 

![Screenshot_20200228_185422](https://user-images.githubusercontent.com/16562333/75572796-c90fc880-5a5b-11ea-9750-37a9afe57748.png)

stephanlachnit replied:
What about renaming it to UNKOWN?

K1ngjulien replied:
hmm yeah, or maybe `INVALID` ?

stephanlachnit replied:
Also possible, but I think `INVALID` implies an error, which it doesn't really do.

FFY00 commented on src/libratbag-data.c:
```diff
@@ -384,7 +392,7 @@ match(const struct input_id *id, char **strv)
 
 static bool
-file_data_matches(struct ratbag *ratbag,
+parse_device_file(struct ratbag *ratbag,
```
Keep the name.

K1ngjulien replied:
but its so cryptic :(

FFY00 replied:
If you want to change it please do it in a separate PR :)

K1ngjulien replied:
ok :D

FFY00 commented on src/libratbag-data.c:
```diff
@@ -442,6 +450,29 @@ file_data_matches(struct ratbag *ratbag,
 	}
 
+	data->type = g_key_file_get_string(keyfile, GROUP_DEVICE, "DeviceType", NULL);
+	log_info(ratbag, "Device Type: '%s'\n", data->type == NULL ? "NONE" : data->type);
```
Debug

FFY00 commented on src/libratbag-data.c:
```diff
@@ -442,6 +450,29 @@ file_data_matches(struct ratbag *ratbag,
+		log_info(ratbag, "No DeviceType, assuming Mouse in '%s'\n",  basename(path));
+		data->devicetype = MOUSE;
+	}
+	else {
```
`else` in the same line as the bracket.
```
} else {
```
Same bellow.

FFY00 commented on src/libratbag-data.c:
```diff
@@ -499,15 +530,14 @@ ratbag_device_data_new_for_id(struct ratbag *ratbag, const struct input_id *id)
-			goto out;
-		if (file_data_matches(ratbag, file, id, &data))
-			goto out;
+			break;
```
Keep the goto.

K1ngjulien replied:
may I ask why? :D we're not writing assembly

K1ngjulien commented:
> We also need to export the property via DBus so that we can fetch it in piper.

Yeah, I don't know how to do that ^^, currently in the process of finding out how

FFY00 commented:
> Yeah, I don't know how to do that ^^, currently in the process of finding out how

I don't have much time to spend right now but we already do that for other properties so have a look at the code. @stephanlachnit might be able to help.

K1ngjulien commented on src/libratbag-data.c:
```diff
@@ -442,6 +450,26 @@ file_data_matches(struct ratbag *ratbag,
+	} else {
+		if(streq(data->type, "keyboard")) {
+			data->devicetype = TYPE_KEYBOARD;
+		}else if (streq(data->type, "mouse")) {
```
whoops, missed a space :man_facepalming:

stephanlachnit commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+builddir/
```
Please drop the `.gitignore`. You can add a local one, meaning it looks like this
```gitignore
.gitignore
builddir/
.vscode/
```

stephanlachnit commented on src/libratbag-data.c:
```diff
@@ -53,6 +53,12 @@ enum driver {
 	STEELSERIES,
 };
 
+enum ratbag_device_type {
```
Could you rename it just to `device_type`?

stephanlachnit commented on src/libratbag-data.c:
```diff
@@ -83,8 +89,10 @@ struct ratbag_device_data {
 	int refcount;
 	char *name;
 	char *driver;
+	char *type;
```
maybe `*device_type` to make it a little bit more clear :)

K1ngjulien replied:
should I keep the enum as `devicetype` and the string as `device_type`? been thinking about this for a while

stephanlachnit replied:
I think that should be fine, even if it's not super beatiful

stephanlachnit commented on src/libratbag-data.c:
```diff
@@ -442,6 +450,26 @@ file_data_matches(struct ratbag *ratbag,
+	log_debug(ratbag, "Device Type: '%s'\n", data->type == NULL ? "no type found" : data->type);
+
+	if(data->type == NULL) {
+		// Good idea? Or just add a DeviceType Property to everything
```
I think it's fine for now, eventually we don't want this since we have control over all device files.

stephanlachnit commented on src/libratbag-data.c:
```diff
@@ -442,6 +450,26 @@ file_data_matches(struct ratbag *ratbag,
+		log_debug(ratbag, "No DeviceType, assuming Mouse in '%s'\n",  basename(path));
+		data->devicetype = TYPE_MOUSE;
+	} else {
+		if(streq(data->type, "keyboard")) {
```
Missing space in front of the `if`

stephanlachnit commented on src/libratbag-data.c:
```diff
@@ -442,6 +450,26 @@ file_data_matches(struct ratbag *ratbag,
+			data->devicetype = TYPE_KEYBOARD;
+		}else if (streq(data->type, "mouse")) {
+			data->devicetype = TYPE_MOUSE;
+		} else {
```
you could also just move all devices without a DeviceType here instead of assuming it's a mouse

K1ngjulien replied:
i get a segfault when passing NULL to streq.

stephanlachnit replied:
Ah, that's unlucky. Then nevermind.

stephanlachnit commented:
So first you need to add `device_type` to libratbag's device struct, you find it [here](https://github.com/libratbag/libratbag/blob/master/src/libratbag-private.h#L131).
Once that is in place, you need to add a DBus call to ask for the DeviceType. You need to implement a function in [ratbagd-device.c](https://github.com/libratbag/libratbag/blob/master/ratbagd/ratbagd-device.c#L117) (and probably something similar to [this](https://github.com/libratbag/libratbag/blob/master/src/libratbag.h#L424)) first.
Then you need to add this function to the [vtable](https://github.com/libratbag/libratbag/blob/master/ratbagd/ratbagd-device.c#L223). If you are at this point, the hardest part is over.
The last thing is to sync up the python wrappers, adding entries [here](https://github.com/libratbag/libratbag/blob/master/tools/ratbagd.py#L301) and [here](https://github.com/libratbag/libratbag/blob/master/tools/ratbagc.py.in#L225).
If I didn't forget anything, that's it. You can take a look at https://github.com/libratbag/libratbag/pull/870, it also adds a property (although not in the .device file) and exposes it via DBus, so you should fine some similarities. Don't worry if you need tons of revisions, I needed them as well.

K1ngjulien commented:
It seems to work, I've put some print statements in Piper which get `DeviceType` from the DBus.
I'll look into writing some tests now

stephanlachnit commented:
Don't worry about the tests as long as it works for now. I can review your PR tmw.

K1ngjulien commented:
should I add the DeviceType Property to all .device files? Then I can remove the optionality of that field from the parser and clean up the code a bit.

stephanlachnit commented:
> should I add the DeviceType Property to all .device files? Then I can remove the optionality of that field from the parser and clean up the code a bit.

Yes, thanks. Btw you should rebase your branch. In case you don't know how it works, you can either google it or (since I saw you use vscode) download the gitlens extension and rebase from there in the commit histroy. You basically want all your commits to be fixups, be aware that you have to force push (`git push your_remote_name -f`).

stephanlachnit commented on ratbagd/ratbagd-device.c:
```diff
@@ -137,6 +137,30 @@ static int ratbagd_device_get_device_name(sd_bus *bus,
 }
 
+static int ratbagd_device_get_device_type(sd_bus *bus,
+					  const char *path,
```
Wrong indent. Use the .editorconfig file (there is a vscode extension :]).

K1ngjulien replied:
are you sure? its the same on all other function definitions. also it looks fine in vscode, its only off on github...

![Screenshot_20200301_105418](https://user-images.githubusercontent.com/16562333/75623525-09ca2780-5bab-11ea-8c05-26f12a4ce587.png)

stephanlachnit replied:
Sorry, you're right. I thought github would use editorconfig, so nvm.

whot replied:
it's an issue with rendering diffs, not with whether you use editorconfig. Specifically, the `+ ` at the beginning likes to upset things so it's quite common that diffs aren't correctly aligned even though the code is.

stephanlachnit commented on ratbagd/ratbagd-device.c:
```diff
@@ -195,7 +219,7 @@ static int ratbagd_device_commit(sd_bus_message *m,
 
 static int
-ratbag_device_get_model(sd_bus *bus,
+ratbagd_device_get_model(sd_bus *bus,
```
Pls do this in a separate commit. If you rebase, do `git reset HEAD^`, and only commit this line (click the blue bar left of the line in vscode to stage only one line).

stephanlachnit replied:
@K1ngjulien thanks for the separate commit, can you also adjust the indent in that commit? Looks like 2 spaces to much.

K1ngjulien replied:
should be fixed now

stephanlachnit commented on ratbagd/ratbagd-device.c:
```diff
@@ -222,7 +246,8 @@ ratbag_device_get_model(sd_bus *bus,
 const sd_bus_vtable ratbagd_device_vtable[] = {
 	SD_BUS_VTABLE_START(0),
-	SD_BUS_PROPERTY("Model", "s", ratbag_device_get_model, 0, SD_BUS_VTABLE_PROPERTY_CONST),
+	SD_BUS_PROPERTY("Model", "s", ratbagd_device_get_model, 0, SD_BUS_VTABLE_PROPERTY_CONST),
```
see above

stephanlachnit replied:
@K1ngjulien if you look at the commit history you did a minor mistake - in your second commit, you renamed it from `Model` to `DeviceType` and added `Model` again in the third commit, please change that.

K1ngjulien replied:
yeah, sorry

K1ngjulien replied:
rebase did it again.... i'll try and fix it again.....

stephanlachnit commented on src/libratbag-data.c:
```diff
@@ -539,7 +547,7 @@ ratbag_device_data_new_for_id(struct ratbag *ratbag, const struct input_id *id)
 		free(files[nfiles]);
 	free(files);
-
+	
```
trailing tab

stephanlachnit commented on tools/ratbagc.py.in:
```diff
@@ -262,6 +262,12 @@ class RatbagdDevice(metaclass=MetaRatbag):
         return libratbag.ratbag_device_get_name(self._device)
 
+    @property
+    def devicetype(self):
```
I'd rather call it `device_type`

stephanlachnit commented on tools/ratbagd.py:
```diff
@@ -347,6 +347,11 @@ def name(self):
         return self._get_dbus_property("Name")
 
+    @GObject.Property
+    def devicetype(self):
```
same

stephanlachnit commented:
See comments, but except that lgtm. Please rebase your branch and squash the commits for the new property (other stuff in a separate commit). I don't have time today to test it, probably not tomorrow, but next week I should find time to take a deeper look :)
But nice job, should be merge-able pretty soon.

stephanlachnit commented on src/libratbag-data.c:
```diff
@@ -504,14 +540,14 @@ ratbag_device_data_new_for_id(struct ratbag *ratbag, const struct input_id *id)
 	}
 
-	if (!data)
+	if (data == NULL)
```
haven't tested this, so can you explain why we need to change this?

K1ngjulien replied:
Sorry, force of habit. We don't need to change this because it does the same thing, but I find it more expressive.
If i see ! somewhere I think of booleans, but this is a pointer.

I can change it back if you want.

stephanlachnit replied:
Either that, or do it in separate commit. Just keep the commit history clear, so that we can trace back any changes properly.

K1ngjulien commented:
ok i think I've figured out rebase now and the commit history looks clean. I dropped the `data == NULL` thing since it doesn't really change anything.

I'm not sure tho what you meant with the missing space, as the indentation is the same as all the other functions since I just copied it from there.

I hope I didn't miss anything :)

K1ngjulien commented:
the next commit should probably be something like 'made DeviceType a required property' where I remove the optional if statement and add the property to all .device files.

stephanlachnit commented:
Okay, pretty neat. Was able to run it. Two more things: you need to add the `DeviceType` property to the test device, else you will get `error: testdevice0: failed to fetch device type`.
Also, please print the device type when running `ratbagctl device-name info`.

Btw, in case you didn't know: you can run `sudo ./ratbagd.devel --verbose=debug` in one console to start the compiled daemon, and `sudo ./ratbagctl.devel -vv --use-existing-ratbagd device-name info` in another terminal, without the need to overwrite your system installation.

I would squash the first two commits, and maybe consider using signed-off commits (can be enabled by default in vscode).

stephanlachnit commented:
Ah and btw don't forget to bump the dbus api version in the meson.build file.

K1ngjulien commented:
>  you need to add the `DeviceType` property to the test device, else you will get `error: testdevice0: failed to fetch device type`.

can you help me out here? I dont really know what to add here. just a char * with the type? or just the enum?

K1ngjulien commented:
also I changed all the .device files to now include the device type and hope I didn't make a mistake. I only found the logitech g815, g910 and g915 as keyboards. are there more that I missed?

stephanlachnit commented:
> can you help me out here? I dont really know what to add here. just a char * with the type? or just the enum?

Add `device->DeviceType = "mouse";` to the [`ratbag_new_test_device`](https://github.com/libratbag/libratbag/blob/master/src/libratbag-test.c#L51) function, that should solve it.

> also I changed all the .device files to now include the device type and hope I didn't make a mistake. I only found the logitech g815, g910 and g915 as keyboards. are there more that I missed?

That's all of them IIRC.

stephanlachnit commented on data/devices/data-parse-test.py:
```diff
@@ -77,11 +81,14 @@ def check_section_device(section):
     except KeyError:
         pass
 
+
```
Empty line added

stephanlachnit commented on data/devices/logitech-g903.device:
```diff
@@ -1,4 +1,5 @@
 [Device]
 Name=Logitech G903
 DeviceMatch=usb:046d:c086;usb:046d:4067
+DeviceType=mouse
```
nitpick: you always did the `Driver` property before the `DeviceType` property, so just do it everywhere to stay consistent.

stephanlachnit commented on ratbagd/ratbagd-device.c:
```diff
@@ -137,6 +137,30 @@ static int ratbagd_device_get_device_name(sd_bus *bus,
+	return 0;
+}
+
+
```
One empty line too much

stephanlachnit commented on tools/ratbagctl.in.in:
```diff
@@ -205,9 +205,9 @@ def print_profile(d, p, level):
 
 def print_device(d, level):
     p = d.profiles[0]  # there should be always one
-
```
removed line

K1ngjulien commented:
soooo the tests still fail ^^.
without adding the `device->device_type = "mouse";` the tests run fine on my machine, but they fail once it add it.

K1ngjulien commented:
excuse my naivety but I have literally never written any tests before :see_no_evil:

stephanlachnit commented:
> Are you sure we need that?

Yes, but you put it in the wrong spot, the daemon doesn't even start (segfault). You tried to set the device type _before_ the device is initialized, you need to move the line after the `device = ratbag_device_new(ratbag, NULL, "Test device", &id);` line.

With this it works fine for me, no segfaults, no warnings when starting the daemon.

K1ngjulien commented:
So the function looks like this now:

```C	
        ...
        struct input_id id = {
		.bustype = 0x00,
		.vendor = 0x00,
		.product = 0x00,
		.version = 0x00,
	};

	ratbag_register_test_drivers(ratbag);

	if (getenv("RATBAG_TEST") == NULL) {
		fprintf(stderr, "RATBAG_TEST environment variable not set\n");
		abort();
	}

	device = ratbag_device_new(ratbag, NULL, "Test device", &id);

	device->device_type = "mouse";

	if (!ratbag_assign_driver(device, &device->ids, test_device)) {
		ratbag_device_destroy(device);
		return NULL;
	}
        ...
```

When I look at the logs in `builddir/meson-logs/testlog.txt` I still see this "error message". Is it fine this way?

```
5/8 test-device                             OK       0.04 s 

--- command ---
13:22:47 /home/julian/Documents/code/libratbag/builddir/test-device
--- stdout ---
Running suite(s): device
100%: Checks: 21, Failures: 0, Errors: 0
--- stderr ---
ratbag error: libratbag bug: Test device: invalid number of profiles (0)
ratbag error: libratbag bug: Test device: multiple active profiles
ratbag error: client bug: Requested invalid profile 3
ratbag error: client bug: Requested invalid profile 4
ratbag error: client bug: Requested invalid profile -1
ratbag error: client bug: Requested invalid profile 2147483647
ratbag error: client bug: Requested invalid profile -1
ratbag error: libratbag bug: Test device: invalid report rate list
-------
```

stephanlachnit commented:
> When I look at the logs in builddir/meson-logs/testlog.txt I still see this "error message". Is it fine this way?

I think the output is fine as long as is doesn't exit with an error code, @bentiss?

> Running meson test in builddir is the correct command, right?

Either do `meson test` in `builddir` or `meson builddir && ninja -C builddir test`, both is equivalent afaik.

K1ngjulien commented:
alright then this should work now

whot commented on ratbagd/ratbagd-device.c:
```diff
@@ -223,6 +246,7 @@ ratbag_device_get_model(sd_bus *bus,
 const sd_bus_vtable ratbagd_device_vtable[] = {
 	SD_BUS_VTABLE_START(0),
 	SD_BUS_PROPERTY("Model", "s", ratbag_device_get_model, 0, SD_BUS_VTABLE_PROPERTY_CONST),
+	SD_BUS_PROPERTY("DeviceType", "s", ratbagd_device_get_device_type, 0, SD_BUS_VTABLE_PROPERTY_CONST),
```
In general, string types should be used for something that's either free-form (like the model) or user-exposed (like the name). Where we have an enum, we should use unsigned ints instead - not having to write string parsing code is definitely a bonus :)

As a general rule: in C you avoid strings at all costs wherever you can, because they require too much work.

K1ngjulien replied:
ok but should we still send the strings "keyboard" or "mouse" over the DBus? or just send an int to represent the enum?

stephanlachnit replied:
Just the int, implement the names in Python.

whot commented on src/libratbag-data.c:
```diff
@@ -83,8 +89,10 @@ struct ratbag_device_data {
 	int refcount;
 	char *name;
 	char *driver;
+	char *device_type;
```
duplicate and risky. the enum carries all the information we want, there's no need to have the extra `char *`

K1ngjulien replied:
ok I just copied the style of how driver is implemented

whot commented on src/libratbag-data.c:
```diff
@@ -272,6 +280,13 @@ ratbag_device_data_get_name(const struct ratbag_device_data *data)
+const char *
+ratbag_device_data_get_device_type(const struct ratbag_device_data *data)
+{
+	return data->device_type;
```
if we'd need this (see my comment about enums on the dbus interface though) we can just do this as

```c
if (type == TYPE_MOUSE) return "mouse";
```

whot commented on src/libratbag.c:
```diff
@@ -196,6 +196,11 @@ ratbag_device_new(struct ratbag *ratbag, struct udev_device *udev_device,
+
+	if (device->data != NULL) {
+		device->device_type = (char*)ratbag_device_data_get_device_type(device->data);
+	}
```
nitpick: no `{}` needed around single-line statements.

whot commented on src/libratbag.c:
```diff
@@ -196,6 +196,11 @@ ratbag_device_new(struct ratbag *ratbag, struct udev_device *udev_device,
 	device->data = ratbag_device_data_new_for_id(ratbag, id);
+
+	if (device->data != NULL) {
+		device->device_type = (char*)ratbag_device_data_get_device_type(device->data);
```
There are very few reasons why you'd want to cast a `const char*` into a `char*` and I don't think any of them apply here. As a general rule, if you need to do this, something isn't right.

whot commented on src/libratbag.h:
```diff
@@ -430,6 +430,15 @@ ratbag_device_get_user_data(const struct ratbag_device *device);
+ * @return "mouse" or "keyboard"
+ */
+const char *
+ratbag_device_get_device_type(const struct ratbag_device* device);
```
as already said above, much easier to make the enum a public one and return a limited set here. This also helps with things like internationalization, `TYPE_MOUSE` is always that, but the string "mouse" is only that in english.

FFY00 commented on data/devices/logitech-g-powerplay.device:
```diff
@@ -3,6 +3,7 @@ Name=Logitech G Powerplay
 DeviceMatch=usb:046d:405f
 LedTypes=logo
 Driver=hidpp20
+DeviceType=mouse
```
Create a new type, `Other`, and use it here. Powerplay is a mat.

K1ngjulien replied:
oh right, i missed that.

K1ngjulien commented:
Ok so from what I understand I should just keep the `enum device_type` and get rid of the char* type entirely. seems like a better solution which removes redundancy.

But what should be sent over the DBus to ratbagctl and piper? Should I just convert that enum into static strings in a big switch statement? Or just send it as an Int and let the receiver sort that out?

I storing it as as and enum and then sending a string would be better IMO because that means you'd only have to change it here when another type is added in the future instead of having to change it in all clients as well.

stephanlachnit commented:
> Or just send it as an Int and let the receiver sort that out?

Yes, the receiver doesn't really has to sort it out since we can implement it in `ratbagd.py`

whot commented:
fwiw, the [dbus api design guidelines](https://dbus.freedesktop.org/doc/dbus-api-design.html) are useful to read in this context.

If you still want to send it as string, I'll insist that you'll have to send  "鼠" and "صفحه کلید" just to drive home the point that things only look easy when you force everyone to english :)

Disclaimer: I think this is chinese for mouse and persian for keyboard, but I'm relying on the internet here so apologies for... well, anything ;)

stephanlachnit commented:
@K1ngjulien are you still working on this? It's almost finished, and since I don't have much to do right now I would do the last bits if you don't want to work on this anymore.

K1ngjulien commented:
Yeah sorry I really don't have time right now. Should I give you collaborator access to my repository?

K1ngjulien commented:
ok that is checked

K1ngjulien commented:
my finals are done and I have time to work on this again. i could use some assistance tho.

is there anything in other PRs that have gone though that I should be aware of?

K1ngjulien commented:
from what I can gather, the device type should be sent as a sort of enum. is there a special way to send enums or is sending 0 for mouse and 1 for keyboard enough?

stephanlachnit commented:
> my finals are done and I have time to work on this again. i could use some assistance tho.

Just ping me if a question arises, I'm somewhat short on time, but answering question should be fine.

> is there anything in other PRs that have gone though that I should be aware of?

I'm not aware of one.

> Is it correct that I should rebase my development branch to get all the recent commits?

Yes, you should definitely do that.

> edit: I forgot that I implemented it this way:

I just noticed, maybe define them like this:

```C
enum device_type {
	TYPE_OTHER = 0,
	TYPE_MOUSE,
	TYPE_KEYBOARD,
};
```

Where other will be anything that is either not set (yet) or not a mouse or keyboard. I think this is a good solution.

FFY00 commented:
Feel free to ping me too.

And I agree `TYPE_OTHER` is better, all devices have type, we might not know about it, but they have a type :)

K1ngjulien commented:
Hey! I think I got it working. I've added TYPE_UNSPECIFIED and TYPE_OTHER to the enum. When you think about it, all supported devices should never have an UNSPECIFIED type, since that is a required field in the .device file. Therefore I've added 'other' to the tests which would then be used to represent devices such as mousepads or headsets.

The DeviceType property is sent over the DBus as a uint32, which will then have to be interpreted as the different types by Piper.

Lastly, I will rebase all the changes in this commit into one when everything is done.

K1ngjulien commented:
The next step would probably be inside of Piper: Allowing keyboard and other type to edit all available macro keys.
As it is now, you cannot set the first two keys to a custom macro since the software thinks they are LMB and RMB and therefore won't let you reassign them.

stephanlachnit commented on ratbagd/ratbagd-device.c:
```diff
@@ -137,6 +137,29 @@ static int ratbagd_device_get_device_name(sd_bus *bus,
+	if (!devicetype) {
+		log_error("%s: device type unspecified\n",
+			  ratbagd_device_get_sysname(device));
+		devicetype = 0;
```
shouldn't be here `devicetype = TYPE_UNSPECIFIED`? After all, that's why one defines enums in the first place.

K1ngjulien replied:
I agree. currently the enum is inside `libratbag-data.c` which means it cannot be accessed outside of that file.

Can I just put it inside of `libratbag.h`? Then it would be accessable in all the spots you suggested.

stephanlachnit replied:
Maybe put it inside `src/libratbag-enums.h` and include that if necessary.

K1ngjulien replied:
oh that makes sense, i did not see that file. Will do.

stephanlachnit commented on src/libratbag-test.c:
```diff
@@ -71,6 +71,8 @@ ratbag_device_new_test_device(struct ratbag *ratbag,
 
 	device = ratbag_device_new(ratbag, NULL, "Test device", &id);
 
+	device->devicetype = 1; // mouse
```
same here, `device->devicetype = TYPE_MOUSE`

stephanlachnit commented on tools/ratbagc.py.in:
```diff
@@ -262,6 +262,12 @@ class RatbagdDevice(metaclass=MetaRatbag):
 
+    @property
+    def device_type(self):
+        """The type of this device; mouse or keyboard"""
```
Needs to be updated

stephanlachnit commented on src/libratbag.h:
```diff
@@ -430,6 +430,15 @@ ratbag_device_get_user_data(const struct ratbag_device *device);
+ * @ingroup device
+ *
+ * @param device A previously initialized ratbag device
+ * @return "mouse" or "keyboard"
```
needs to be updated

stephanlachnit commented on tools/ratbagd.py:
```diff
@@ -347,6 +347,11 @@ def name(self):
         """The device name, usually provided by the kernel."""
         return self._get_dbus_property("Name")
 
+    @GObject.Property
+    def device_type(self):
+        """The type of this device: unspecified(0), mouse(1), keyboard(2) or other(3)"""
+        return self._get_dbus_property("DeviceType")
+
```
I think it's probably a good idea to add the enums in ratbagd somehow. Maybe like [here](https://github.com/libratbag/libratbag/blob/ceaa3c3ad5c7b4b7ebe1bc5ca8ae1c1a30ecb006/tools/ratbagd.py#L37) (just an idea).

stephanlachnit commented on tools/ratbagctl.in.in:
```diff
@@ -212,6 +212,7 @@ def print_device(d, level):
 
     print(" " * level + "{} - {}".format(d.id, d.name))
     print(" " * level + "             Model: {}".format(d.model))
+    print(" " * level + "       Device Type: {}".format(d.device_type))
```
This needs to be changed, see my other comment.

K1ngjulien commented:
Oh shit something failed on fedora. don't know what that means.

K1ngjulien commented:
oh shit something failed on fedora,  don't know what that means.

all the requested changes should be in now and I'll squash them when you give the OK so that the others have an easier time reviewing the changes.

K1ngjulien commented on src/libratbag.h:
```diff
@@ -434,9 +434,9 @@ ratbag_device_get_name(const struct ratbag_device* device);
  *
  * @param device A previously initialized ratbag device
- * @return "mouse" or "keyboard"
+ * @return the type of of the device: unspecified, mouse, keyboard or other
```
typo...

stephanlachnit commented:
> oh shit something failed on fedora, don't know what that means.

It's just a weird DBus error, happens sometimes, I restarted it.

stephanlachnit commented on src/libratbag-enums.h:
```diff
@@ -327,3 +327,22 @@ enum ratbag_macro_event_type {
+enum ratbag_device_type {
+	// Used when no DeviceType property could be found
+	TYPE_UNSPECIFIED = 0,
+	// Used for mice
+	TYPE_MOUSE,
+	// Used for Keyboards
+	TYPE_KEYBOARD,
```
Just for consistency, if you write "Keyboard" with a capital letter in the beginning, do it for "mice" as well.

K1ngjulien replied:
ok

stephanlachnit commented on tools/ratbagd.py:
```diff
@@ -347,6 +359,11 @@ def name(self):
+    @GObject.Property
+    def device_type(self):
+        """The type of this device: unspecified(0), mouse(1), keyboard(2) or other(3)"""
+        return RatbagDeviceType(self._get_dbus_property("DeviceType"))
+
```
Description doesn't equal return value.

K1ngjulien replied:
What would be better?

`The type of this device as specified in the .device file` ?

or would be `The type of this device` good enough?

stephanlachnit replied:
`The device type, see RatbagDeviceType.` is good enough IMHO. I think pointing out the enum explicitly is a good idea.

stephanlachnit commented on tools/ratbagc.py.in:
```diff
@@ -262,6 +262,12 @@ class RatbagdDevice(metaclass=MetaRatbag):
         """The device name, usually provided by the kernel."""
         return libratbag.ratbag_device_get_name(self._device)
 
+    @property
+    def device_type(self):
+        """The type of this device: unspecified(0), mouse(1), keyboard(2) or other(3)"""
+        return libratbag.ratbag_device_get_device_type(self._device)
+
+
```
The enum should be defined in `ratbagc.py` as well, the behavior of this function should be the same in both `ratbagd.py` and `ratbagc.py`.

K1ngjulien replied:
will fix, did not know that.

K1ngjulien commented:
sorry made a mistake while rebasing. change was in the wrong commit.

stephanlachnit commented on src/libratbag-enums.h:
```diff
@@ -327,3 +327,21 @@ enum ratbag_macro_event_type {
+	TYPE_MOUSE,
+	// Used for keyboards
+	TYPE_KEYBOARD,
+	// Used for any other device type that doesn't need a special type
+	// Headsets, RGB Mousepads etc
```
To make it more consistent with the description in `ratbagd.py`:
```C
	// Used for any other device type than a mouse or a keyboard
	// for example headsets, mousepads etc.
```

stephanlachnit commented on tools/ratbagc.py.in:
```diff
@@ -262,6 +275,12 @@ class RatbagdDevice(metaclass=MetaRatbag):
 
+    @property
+    def device_type(self):
+        """The type of this device provided by a .device file"""
```
Same here.

K1ngjulien commented:
Ok great! Doc comment changes are in, commits from upstream are rebased. 

Implementing this in piper before merging seems like a good idea.

FFY00 commented on ratbagd/ratbagd-device.c:
```diff
@@ -195,13 +218,13 @@ static int ratbagd_device_commit(sd_bus_message *m,
-			   sd_bus_message *reply,
-			   void *userdata,
-			   sd_bus_error *error)
+ratbagd_device_get_model(sd_bus *bus,
```
What is the difference between the model and the name?

K1ngjulien replied:
don't know, I just fixed the name. it used to be without the d at the end, which clashed with another function iirc.

FFY00 commented on src/libratbag-data.c:
```diff
@@ -281,6 +282,13 @@ ratbag_device_data_get_name(const struct ratbag_device_data *data)
 }
 
+enum ratbag_device_type
+ratbag_device_data_get_device_type(const struct ratbag_device_data *data)
```
Do we need this? I think we only use this when we need to have an assert.

K1ngjulien replied:
all other properties have it so i added it here as well

FFY00 commented on src/libratbag-enums.h:
```diff
@@ -327,3 +327,21 @@ enum ratbag_macro_event_type {
+	TYPE_KEYBOARD,
+	// Used for any other device type than a mouse or a keyboard
+	// for example headsets, mousepads etc.
+	TYPE_OTHER
```
nitpick: I would move this up, to the 2nd position. The argument being it is a "special" type and when we add new types we will have a special type in the middle of the list.

Also adjust the documentation after.

K1ngjulien replied:
realistically, mice are the only one needing a special case, because the first two buttons shouldn't be rempped. I don't think that there will be many special cases in the future.

FFY00 replied:
Well yes, but I would still like to keep the enum in order. Should be a pretty easy fix.

stephanlachnit replied:
I think it is a good idea as well

K1ngjulien replied:
its swapped now

K1ngjulien commented:
pullrequest to use these changes in piper

https://github.com/libratbag/piper/pull/536

stephanlachnit commented on ratbagd/ratbagd-device.c:
```diff
@@ -137,6 +137,29 @@ static int ratbagd_device_get_device_name(sd_bus *bus,
+	if (!devicetype) {
+		log_error("%s: device type unspecified\n",
+			  ratbagd_device_get_sysname(device));
+		devicetype = TYPE_MOUSE;
```
This is incorrect, it needs to be `TYPE_UNSPECIFIED`.

stephanlachnit commented on src/libratbag-data.c:
```diff
@@ -451,6 +460,28 @@ file_data_matches(struct ratbag *ratbag,
+	log_debug(ratbag, "Device Type: '%s'\n", devicetype == NULL ? "no type found" : devicetype);
+
+	if (devicetype == NULL) {
+		log_error(ratbag, "No DeviceType, found in '%s'\n",  basename(path));
```
A comma too much, i. e. `"No DeviceType found in '%s'\n"`.

stephanlachnit commented:
@whot do you have time to give this and https://github.com/libratbag/piper/pull/536 a quick review?
I think it's basically ready to merge (except some really minor things), if you give your green light and the minor things get fixed, we can finally merge this.

K1ngjulien commented:
minor things have been fixed

whot commented on data/devices/data-parse-test.py:
```diff
@@ -55,6 +55,10 @@ def check_match_str(string):
         pid = parts[2]
         assert(pid == '{:04x}'.format(int(pid, 16)))
```
adding this here because I can't comment on the commit message in the review interface: please use present tense for commit messages. Think of the patch doing something, i.e. "add foo to bar"  (not ~~added foo to bar~~)

K1ngjulien replied:
ok I'll change the name in the rebase

whot commented on src/libratbag-enums.h:
```diff
@@ -327,3 +327,21 @@ enum ratbag_macro_event_type {
+ * Describes the types specified in the .device files
+ */
+enum ratbag_device_type {
+	// Used when no DeviceType property could be found
```
`/* */` comments please

K1ngjulien replied:
will do

whot commented on src/libratbag-enums.h:
```diff
@@ -327,3 +327,21 @@ enum ratbag_macro_event_type {
+	TYPE_UNSPECIFIED = 0,
+	// Used for any other device type than a mouse or a keyboard
+	// for example headsets, mousepads etc.
+	TYPE_OTHER,
```
Generally I recommend against a type "other" because it will eventually break things. Probably not a big deal for libratbag since no-one cares about minor breaks here but if, eventually, you want to add a new type `TYPE_HEADSET` then any consumer that relies on the type being `OTHER` will break.

Given that we don't appear to support any "other" device anyway I think it's better to just leave this out (of the API as well) and explicitly don't support anything that's not a mouse or keyboard. Once you have sufficient requests for other devices you'll also know better how to categorise those.

K1ngjulien replied:
we do use "other", there is the logitech powerplay mousepad and a headset somewhere in the data files to controll their RGB.

if there's a need to add a narrower specification, like headset, then it can be added in the future.
I don't see why code depending on "other", which there currently is none,  cannot be changed at that time.
wouldn't adding a new type need a new api version anyway?

stephanlachnit replied:
I also think having a `OTHER` type is a good idea. I guess it would be also possible to merge `TYPE_OTHER` and `TYPE_UNSPECIFIED`, since in the end it doesn't make that much of a difference thanks to capabilities. But on the other side, the type of a headset isn't really unspecified, it's just not one of the types where the type might be relevant.

> wouldn't adding a new type need a new api version anyway?

Yes it would, since sending an other enum than specified will break in `ratbagd.py`.

whot replied:
>  wouldn't adding a new type need a new api version anyway?

The history of the version field is that we never managed to declare a stable API. The original plan was to have the API stable enough that it won't change for ages and be backwards compatible - because that may be the only way to get something other than piper support ratbagd. After all, the point of libratbag was to have a system service that does the device abstraction without being tied to any specific UI. Never managed to get to that point because ETIME. The version field was introduced to stop piper from crashing when something in ratbagd changed, it was never intended to be a long-term solution, only a hack until the API was stable enough. A decent version negotiation would allow for backwards compatibility, this wasn't planned for because declaring the API stable was always just around the corner.

stephanlachnit replied:
I guess when we're at the point that we can declare a stable API, we could remove `TYPE_OTHER` and categorise everything we have until then. But it would be a bummer if we have to bump for every new device that isn't a mouse or a keyboard the API. For example, maybe there will be a gamepad, or a display with RGB, etc.

K1ngjulien replied:
so should I change anything here? fact is we need this api change to enable differentiating between device types.

whot replied:
I don't do enough work on libratbag these days to veto this, so feel free to merge with this.

Might be worth thinking about capabilities instead of types - i.e. does the device have keys, mouse-like axes, LEDs etc. Provides a more generic interface because if what we want to configure is a headset's LEDs we don't need to care that it's a headset. Likewise we don't care that something is a mouse, we care about buttons, resolutions and LEDs - the only thing that hints at a mouse is the rendering in piper but that relies on the SVG anyway.

stephanlachnit replied:
> Might be worth thinking about capabilities instead of types - i.e. does the device have keys, mouse-like axes, LEDs etc. Provides a more generic interface because if what we want to configure is a headset's LEDs we don't need to care that it's a headset. Likewise we don't care that something is a mouse, we care about buttons, resolutions and LEDs - the only thing that hints at a mouse is the rendering in piper but that relies on the SVG anyway.

Actually that's a good idea -  I can open an issue and put it on the roadmap.

> I don't do enough work on libratbag these days to veto this, so feel free to merge with this.

I would be fine merging this then, @FFY00 if you are fine with as well can you merge this and https://github.com/libratbag/piper/pull/536?

K1ngjulien commented:
@whot any comments on this?  can this be merged?

FFY00 commented:
Please no merge commits if possible.

staticssleever668 commented on ratbagd/ratbagd-device.c:
```diff
@@ -137,6 +137,29 @@ static int ratbagd_device_get_device_name(sd_bus *bus,
+	if (!devicetype) {
+		log_error("%s: device type unspecified\n",
+			  ratbagd_device_get_sysname(device));
+		devicetype = TYPE_UNSPECIFIED;
```
This is a no-op because `!devicetype` is equal to `devicetype == 0` and `TYPE_UNSPECIFIED` is `0`.

staticssleever668 commented on src/libratbag-data.c:
```diff
@@ -403,6 +411,7 @@ file_data_matches(struct ratbag *ratbag,
 	_cleanup_(g_strfreevp) char **ledtypes_strv = NULL;
 	_cleanup_(ratbag_device_data_unrefp) struct ratbag_device_data *data = NULL;
 	int rc;
+	char *devicetype;
```
Needs a `_cleanup_free_`.

staticssleever668 replied:
..or it doesn't? This causes a segmentation fault. Sorry.

Although `g_key_file_get_string()` is supposed to return a newly allocated string per glib docs, perhaps there is more into it.

whot replied:
did you init it to `NULL`? There are some paths where a `_cleanup_free_` would be triggered and cause a free of an uninitialized value - maybe that was the reason? glib definitely returns an allocated string here.

staticssleever668 replied:
Oh, you are right, I missed that!

staticssleever668 commented on tools/ratbagctl.in.in:
```diff
@@ -229,6 +229,7 @@ def print_device(d, level):
 
     print(" " * level + "{} - {}".format(d.id, d.name))
     print(" " * level + "             Model: {}".format(d.model))
+    print(" " * level + "       Device Type: {}".format(d.device_type.name))
```
TODO: find a prettier way to format this, because right now this shouts "MOUSE".

whot replied:
Python's `str.capitalize()` will make this less agressive :)

staticssleever668 replied:
Thanks!

staticssleever668 commented:
Rebased adding DeviceType to all of the new devices. Addressed my nit picks in separate fixup commits.

staticssleever668 commented on src/libratbag-data.c:
```diff
@@ -449,6 +450,13 @@ ratbag_device_data_get_name(const struct ratbag_device_data *data)
+	return data->devicetype;
+}
+
+
```
nit: unneeded new line

staticssleever668 commented:
Sorry for the ton of force pushes, I only keep noticing style issue after pushing...
(Would be nice to just have automatic formatting not to think about all of this).

staticssleever668 commented:
Thank you!

K1ngjulien commented:
🥳🥳🥳
